src/array_index.jl

```diff
@@ -288,3 +406,126 @@ end
+end
+@inline function Base.getindex(x::StrideIndex{N,R,C}, ::PermutedIndex{N,perm,iperm}) where {N,R,C,perm,iperm}
+    if C === nothing || C === -1
+        c2 = C
+    else
+        c2 = getfield(iperm, C)
+    end
+    return StrideIndex{N,permute(R, Val(perm)),c2}(
+        permute(strides(x), Val(perm)),
+        permute(offsets(x), Val(perm)),
+    )
+end
+@inline function Base.getindex(x::PermutedIndex, i::PermutedIndex)
+    PermutedIndex(
+        permute(to_parent_dims(x), to_parent_dims(i)),
+        permute(from_parent_dims(x), from_parent_dims(i))
+    )
+end
+
+@inline function Base.getindex(x::LinearSubIndex, i::LinearSubIndex)
+    s = getfield(x, :stride)
+    LinearSubIndex(
+        getfield(x, :offset) + getfield(i, :offset) * s,
+        getfield(i, :stride) * s
+    )
+end
+Base.getindex(::OffsetIndex{StaticInt{0}}, i::StrideIndex) = i
+
+
+## ArrayIndex constructorrs
+@inline _to_cartesian(a) = CartesianIndices(ntuple(dim -> indices(a, dim), Val(ndims(a))))
+@inline function _to_linear(a)
+    N = ndims(a)
+    StrideIndex{N,ntuple(+, Val(N)),nothing}(size_to_strides(size(a), static(1)), offsets(a))
+end
+
+## DenseArray
+"""
+    ArrayIndex{N}(A) -> index
+
+Constructs a subtype of `ArrayIndex` such that an `N` dimensional indexing argument may be
+converted to an appropriate state for accessing the buffer of `A`. For example:
+
+```julia
+julia> A = reshape(1:20, 4, 5);
+
+julia> index = ArrayInterface.ArrayIndex{2}(A);
+
+julia> ArrayInterface.buffer(A)[index[2, 2]] == A[2, 2]
+true
+
+```
+"""
+ArrayIndex{N}(x::DenseArray) where {N} = StrideIndex(x)
+ArrayIndex{1}(x::DenseArray) = OffsetIndex(static(0))
+
+ArrayIndex{1}(x::ReshapedArray) = OffsetIndex(static(0))
+ArrayIndex{N}(x::ReshapedArray) where {N} = _to_linear(x)
+
+ArrayIndex{1}(x::AbstractRange) = OffsetIndex(static(0))
+
+## SubArray
+ArrayIndex{N}(x::SubArray) where {N} = SubIndex{ndims(x)}(getfield(x, :indices))
+function ArrayIndex{1}(x::SubArray{<:Any,N}) where {N}
+    ComposedIndex(_to_cartesian(x), SubIndex{N}(getfield(x, :indices)))
+end
+ArrayIndex{1}(x::Base.FastContiguousSubArray) = OffsetIndex(getfield(x, :offset1))
+function ArrayIndex{1}(x::Base.FastSubArray)
+    LinearSubIndex(getfield(x, :offset1), getfield(x, :stride1))
+end
+
+## Permuted arrays
+ArrayIndex{2}(::MatAdjTrans) = PermutedIndex{2,(2,1),(2,1)}()
+ArrayIndex{2}(::VecAdjTrans) = PermutedIndex{2,(2,1),(2,)}()
+ArrayIndex{1}(x::MatAdjTrans) = ComposedIndex(_to_cartesian(x), ArrayIndex{2}(x))
+ArrayIndex{1}(x::VecAdjTrans) = OffsetIndex(static(0))  # just unwrap permuting struct
+ArrayIndex{1}(::PermutedDimsArray{<:Any,1}) = OffsetIndex(static(0))
+function ArrayIndex{N}(::PermutedDimsArray{<:Any,N,perm,iperm}) where {N,perm,iperm}
+    PermutedIndex{N,perm,iperm}()
+end
+function ArrayIndex{1}(x::PermutedDimsArray{<:Any,N,perm,iperm}) where {N,perm,iperm}
+    ComposedIndex(_to_cartesian(x), PermutedIndex{N,perm,iperm}())
+end
+
```

Comment:
Because many `SubArray`s and `PermutedDimArray`s can be represented by `StridedIndex`, can we have `ArrayIndex` for these return a `StridedIndex`?

Reply:
The way I'm treating `ArrayIndex` here is a one layer deep index representation. The problem with going straight to `StrideIndex` is that whatever is calling `ArrayIndex(::AbstractArray)` doesn't know how deeply nested the buffer is. I figured the user level interface would be more like `layout(A, accessor) = ArrayIndex(A), buffer(A)` and we could special case for cartesian indexing going straight to `StrideIndex`.

Reply:
@chriselrod, should this PR have something like the `layout` method implemented so that it is immediately more useful? I was trying to make a PR that was still functional but didn't create too many new things at once, but I can see how that lack of a fully implemented interface runs the risk of breaking changes in the near future. Alternatively, I could add warnings to the docs that this is experimental, so that we can drop/change stuff without worrying about breaking changes for now.

Reply:
> Alternatively, I could add warnings to the docs that this is experimental, so that we can drop/change stuff without worrying about breaking changes for now.

Sure, although why not have these things live in a separate, experimental repo?
Are they part of the array interface, or something other libraries would want to extend?

It's hard for me to assess changes like these if I don't know the vision for their future use.

Reply:
Hopefully recent commits involving the indexing pipeline make it more clear that these changes allow unique interactions between layers of nested arrays. If we were to have a separate package for this we'd need ArrayInterface.jl to depend on it so that we could complete the indexing interface.

That being said, we don't need to strictly use what I've proposed here. The point of the `ArrayInterface.compose` methods here is to support a simple rule based system for composing simplified/efficient index transformations. I'm sure will need the new types here, but I'm completely open to there being a more optimal way to doing this.